listing-rewards/program/src/listings/cancel/mod.rs

```diff
@@ -148,24 +151,60 @@ pub fn handler(
-        AuctioneerCancel {
-            wallet: ctx.accounts.wallet.to_account_info(),
-            token_account: ctx.accounts.token_account.to_account_info(),
-            token_mint: ctx.accounts.token_mint.to_account_info(),
-            auction_house: ctx.accounts.auction_house.to_account_info(),
-            auction_house_fee_account: ctx.accounts.auction_house_fee_account.to_account_info(),
-            trade_state: ctx.accounts.trade_state.to_account_info(),
-            authority: ctx.accounts.authority.to_account_info(),
-            auctioneer_authority: ctx.accounts.reward_center.to_account_info(),
-            ah_auctioneer_pda: ctx.accounts.ah_auctioneer_pda.to_account_info(),
-            token_program: ctx.accounts.token_program.to_account_info(),
-        },
-        reward_center_signer_seeds,
-    );
-
-    mpl_auction_house::cpi::auctioneer_cancel(cancel_accounts_ctx, price, token_size)?;
+    let auction_house_program = ctx.accounts.auction_house_program.to_account_info();
+
+    let cancel_listing_ctx_accounts = AuctioneerCancel {
+        wallet: ctx.accounts.wallet.to_account_info(),
+        token_account: ctx.accounts.token_account.to_account_info(),
+        token_mint: ctx.accounts.token_mint.to_account_info(),
+        auction_house: ctx.accounts.auction_house.to_account_info(),
+        auction_house_fee_account: ctx.accounts.auction_house_fee_account.to_account_info(),
+        trade_state: ctx.accounts.trade_state.to_account_info(),
+        authority: ctx.accounts.authority.to_account_info(),
+        auctioneer_authority: ctx.accounts.reward_center.to_account_info(),
+        ah_auctioneer_pda: ctx.accounts.ah_auctioneer_pda.to_account_info(),
+        token_program: ctx.accounts.token_program.to_account_info(),
+    };
+
+    let cancel_listing_params = AuctioneerCancelParams {
+        buyer_price: price,
+        token_size,
+    };
+
+    let auth_account_key = if auction_house.requires_sign_off {
+        auction_house_authority_key
+    } else {
+        wallet_key
+    };
+
+    let signer_required_keys = vec![ctx.accounts.reward_center.key(), auth_account_key];
+
+    let cancel_listing_ix = Instruction {
+        program_id: auction_house_program.key(),
+        data: cancel_listing_params.data(),
+        accounts: cancel_listing_ctx_accounts
+            .to_account_metas(None)
+            .into_iter()
+            .map(|mut account| {
+                if signer_required_keys.contains(&account.pubkey) {
+                    account.is_signer = if account.pubkey.eq(&reward_center.key()) {
+                        true
+                    } else if account.pubkey.eq(&wallet_key) {
+                        ctx.accounts.wallet.to_account_info().is_signer
+                    } else {
+                        ctx.accounts.authority.to_account_info().is_signer
+                    }
+                }
+                account
+            })
+            .collect(),
```

Comment:
This seems to have gotten more complicated and rigid than the zip implementation showcased in auctioneer.